Pid thermostat (#27382)
home-assistant/home-assistant.github.io: source/_integrations/pid_thermostat.markdown

---
title: PID Thermostat
description: Creates a PID thermostat to regulate the temperature in a room with PID algorithm.
ha_category:
  - Number
ha_release: '2023.05'
ha_iot_class: Local Polling
ha_domain: pid_thermostat
ha_platforms:
  - number
ha_integration_type: integration
---

> The `pid_thermostat` climate platform is a thermostat implemented in Home Assistant. It uses a sensor and a number entity connected to a heater or air conditioning under the hood. A typical method to create a number to control a heater or cooler could be to use the [slow_pwm number integration][slow_pwm] . When in heater mode, if the measured temperature is cooler than the target temperature, the heater will be regulated to the required temperature is reached. When in air conditioning mode, if the measured temperature is hotter than the target temperature, the air conditioning will be regulated to the required temperature. One Generic Thermostat entity can only control one number output. If you need to activate two numbers, one for a heater and one for an air conditioner, you will need two Generic Thermostat entities. 

Reviewer: ```suggestion
The `pid_thermostat` climate platform is a thermostat implemented in Home Assistant. It uses a sensor and a number entity connected to a heater or air conditioning under the hood. A typical method to create a number to control a heater or cooler could be to use the [slow_pwm number integration][slow_pwm]. When in heater mode, if the measured temperature is cooler than the target temperature, the heater will be regulated to the required temperature is reached. When in air conditioning mode, if the measured temperature is hotter than the target temperature, the air conditioning will be regulated to the required temperature. One Generic Thermostat entity can only control one number output. If you need to activate two numbers, one for a heater and one for an air conditioner, you will need two Generic Thermostat entities. 
```

## Configuration
{% include integrations/config_flow.md %}
{% configuration_basic %}
Name:
  description: The name that the PID thermostat should have. You can change it again later.
Heater device entity:
  description: Heater- or cooler number entity. Typically, the [slow_pwm number entity][slow_pwm] can be used to create a number controlling a switch.
Heater device entity:
  description: Temperature sensor, used for input signal.
Proportional gain factor (Kp):
  description: Proportional gain factor, directly gaining the error to compensate the fault (Kp).
Integration factor (Ki):
  description: Integration factor, reducing offset fault over time (Ki).
Differential factor (Kd):
  description: Differential factor, damping the overshoot (Kd).
Thermostat mode: 
  description: Select if the thermostat should be a cooler or a heater.
Duration between controller cycles
  description: Cycle time that elapses between two controller cycles.
{% endconfiguration_basic %}

## YAML Configuration

> Alternatlively, this integration can be configured and set up manually via YAML

Reviewer: ```suggestion
Alternatively, this integration can be configured and set up manually via YAML.
```

> instead. To enable the PID thermostat entity in your installation, add the

Reviewer: ```suggestion
To enable the PID thermostat entity in your installation, add the
```

following to your `configuration.yaml` file:

```yaml
# Example configuration.yaml entry
climate:
  - platform: pid_thermostat
    name: Kitchen thermostat
    heater: number.floor_heater
    sensor: sensor.kitchen_temperature
``` 

{% configuration %}
name:
  description: Name of the PID thermostat.
  required: true
  type: string
heater:

>   description: `entity_id` for heater- or cooler, must be number device. Typically, the [slow_pwm number entity][slow_pwm] can be used to create a number controlling a switch.

Reviewer: ```suggestion
  description: `entity_id` for heater or cooler. Must be a number device. Typically, the [slow_pwm number entity][slow_pwm] can be used to create a number controlling a switch.
```

  required: true
  type: string
sensor:
  description: `entity_id` for temperature sensor.
  required: true
  type: string
kp:
  description: Proportional gain factor, directly gaining the error to compensate the fault (Kp).
  required: false
  default: 100.0
  type: float
ki:

>   description: Integration factor, reducing offset fault over time (Ki).

Reviewer: ```suggestion
  description: Integration factor, reducing the offset fault over time (Ki).
```

  required: false
  default: 0.1
  type: float
kd:
  description: Differential factor, damping the overshoot (Kd).
  required: false
  default: 0.0
  type: float
ac_mode:
  description: Thermostat mode. 
  required: false
  default: direct
  type: string ('cool' or 'heat')
min_temp:

>   description: Minimal temperature setpoint.

Reviewer: ```suggestion
  description: Minimal temperature setpoint in °C.
```
I'm assuming Celsius is used here?

  required: false
  default: 7
  type: float
max_temp:

>   description: Maximal temperature setpoint.

Reviewer: ```suggestion
  description: Maximal temperature setpoint in °C.
```

  required: false
  default: 35
  type: float
cycle_time:
  description: Cycle time for the internal PID controller.
  required: false
  default: {"seconds": 30}
  type: time_period
target_temp:

>   description: Target tempererature on startup.

Reviewer: ```suggestion
  description: Target temperature on startup.
```

  required: false
  default: 19
  type: float
initial_hvac_mode:

>   description: Initial hvac mode. 

Reviewer: ```suggestion
  description: Initial hvac mode. 
```
what is hvac?

Author: HVAC stands for heating, ventilation, and air conditioning. I didn't know that abbreviation as well, but it's used by the climate integrations of HASS. Changed it to captions now.

  required: false
  default: off

>   type: string ('off', 'cool' or 'heat')

Reviewer: ```suggestion
  type: string ('off', 'cool', or 'heat')
```

Reviewer: I'm probably not the target audience for this text and maybe this is clear to people who know their way around PID and PWM. How do I know what values to choose for kp, ki, kd? kp defaults to 100, yet in your example you went for 150.

Author: Let me add the same explainer to this text as for pid_contoller. When used in a thermostat, one usually works with high kp; if the temperature is 1 degC lower than wished, the heater should switch on at full power, because that's already relative much for such an application.

The value for the output number entity will be calculated using the Proportional–Integral–Derivative algorithm (PID, See [https://en.wikipedia.org/wiki/PID_controller]). The implementation of the PID controller contains bumpless operation, and is prevented against integral windup by clipping of the output value to the minimum and maximum of the corresponding output number entity. 

> Setting up the optimal parameters for a PID controller can be a though job. Depending on your particalar job, you might already know more or less what the parameters should be. If required, you could use [manual tuning][https://en.wikipedia.org/wiki/PID_controller#Manual_tuning] to find optimal parameters. 

Reviewer: ```suggestion
Setting up the optimal parameters for a PID controller can be a tough job. Depending on your particular job, you might already know more or less what the parameters should be. If required, you could use [manual tuning][https://en.wikipedia.org/wiki/PID_controller#Manual_tuning] to find optimal parameters. 
```

> - For kp, start with 100; if your thermostat deviates 1 °C you might want the heater to turn on for 100%. If required, gradually make it bigger if you see that the direct reaction of the controller is too low.

Reviewer: ```suggestion
- For kp, start with 100; if your thermostat deviates by 1 °C, you might want the heater to turn on for 100%. If required, gradually make it bigger if you see that the direct reaction of the controller is too low.
```

> - For ki, keep this number to 0 until kp is set. Than start with a small number (0.1). If you see that the reaction over time is only slowly rising, than increase it, until the controller regulates to the sepoint in a reasonalble amount of time. 

Reviewer: ```suggestion
- For ki, keep this number to 0 until kp is set. Then, start with a small number (0.1). If you see that the reaction over time is only slowly rising, then increase it, until the controller regulates to the setpoint in a reasonable amount of time. 
```

> - For kd, keep this number to 0 until kp and ki are set. Now you can use the kd to prevent the regulator from overshoot. Only increase in small steps.

Reviewer: ```suggestion
- For kd, keep this number to 0 until kp and ki are set. Now you can use the kd to prevent the regulator from overshooting. Only increase in small steps.
```

name:
  description: Name of the PID thermostat.
  required: true
  type: string
heater:
  description: `entity_id` for heater or cooler. Must be a number device. Typically, the [slow_pwm number entity][slow_pwm] can be used to create a number controlling a switch.
  required: true
  type: string
sensor:
  description: `entity_id` for temperature sensor.
  required: true
  type: string
kp:
  description: Proportional gain factor, directly gaining the error to compensate the fault (Kp).
  required: false
  default: 100.0
  type: float
ki:
  description: Integration factor, reducing the offset fault over time (Ki).
  required: false
  default: 0.1
  type: float
kd:
  description: Differential factor, damping the overshoot (Kd).
  required: false
  default: 0.0
  type: float
ac_mode:
  description: Thermostat mode. 
  required: false
  default: direct
  type: string ('cool' or 'heat')
min_temp:
  description: Minimal temperature setpoint in °C.
  required: false
  default: 7
  type: float
max_temp:
  description: Maximal temperature setpoint in °C.
  required: false
  default: 35
  type: float
cycle_time:
  description: Cycle time for the internal PID controller.
  required: false

>   default: {"seconds": 30}

Reviewer: ```suggestion
  default: "{'seconds': 30}"
```
Not 100% sure about this. Could you check and see if this fixes the pipeline run?
